File: solutions/realtimeRetrieval/RealtimeRetrieval/src/main/scala/com/zr/police/KafkaStreaming.scala
package main.scala.com.zr.police

import java.io.{File, FileInputStream}
import java.util.Properties

import main.scala.com.zr.utils.LoginUtil
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.hbase.{HBaseConfiguration, HColumnDescriptor, HTableDescriptor}
import org.apache.hadoop.hbase.client.{HBaseAdmin, HTable, Put}
import org.apache.hadoop.hbase.io.ImmutableBytesWritable
import org.apache.hadoop.hbase.mapreduce.{TableInputFormat, TableOutputFormat}
import org.apache.hadoop.hbase.util.Bytes
import org.apache.hadoop.mapred.JobConf
import org.apache.spark.streaming.kafka010.{ConsumerStrategies, KafkaUtils, LocationStrategies}
import org.apache.spark.streaming.{Seconds, StreamingContext}
import org.apache.spark.{SparkConf, SparkContext}

object KafkaStreaming {
  def main(args: Array[String]): Unit = {
    //读取配置文件
    val propertie = new Properties()
    val in = this.getClass.getClassLoader().getResourceAsStream("police.properties")
    propertie.load(in)
    //进行安全认证
    val userPrincipal = propertie.getProperty("userPrincipal")
    val userKeytabPath = propertie.getProperty("userKeytabPath")
    val krb5ConfPath = propertie.getProperty("krb5ConfPath")
    val hadoopConf: Configuration = new Configuration()
    LoginUtil.login(userPrincipal, userKeytabPath, krb5ConfPath, hadoopConf)
    //spark配置信息
    val conf = new SparkConf().setMaster("local[2]").setAppName("kafka")
    val ssc = new StreamingContext(conf, Seconds(2))
    ssc.checkpoint(propertie.getProperty("chekPath"))

    val hbaseConf = HBaseConfiguration.create()
    hbaseConf.set("hbase.zookeeper.property.clientPort", propertie.getProperty("port"))
    hbaseConf.set("hbase.zookeeper.quorum", propertie.getProperty("zkquorum"))
    val tableName = propertie.getProperty("tableName")
        hbaseConf.set(TableInputFormat.INPUT_TABLE, tableName)
    val jobConf = new JobConf(hbaseConf)
//    jobConf.setOutputFormat(classOf[TableOutputFormat])
    jobConf.set(TableOutputFormat.OUTPUT_TABLE,tableName)
    val topics = propertie.getProperty("topic").split("\\|").toSet
    val groupId = propertie.getProperty("groupId")
    val brokers = propertie.getProperty("brokers")
    val kafkaParams = Map[String, String](
      "bootstrap.servers" -> brokers,
      "value.deserializer" -> "org.apache.kafka.common.serialization.StringDeserializer",
      "key.deserializer" -> "org.apache.kafka.common.serialization.StringDeserializer",
      "group.id" -> groupId
    )
    /**
      * 酒店住宿：姓名、身份证ID、年龄、性别、入住酒店地址、入住时间、退房时间、同行人。
      * 网吧上网：姓名、身份证ID、年龄、性别、网吧地址、上网日期、时长
      * 卡口身份核验：姓名、身份证ID、年龄、性别、卡口位置、核验时间、出行方式（自驾、乘车、步行）
      */
    val locationStrategy = LocationStrategies.PreferConsistent
    val consumerStrategies = ConsumerStrategies.Subscribe[String, String](topics, kafkaParams)
    val kafkaStream = KafkaUtils.createDirectStream[String, String](ssc, locationStrategy, consumerStrategies)
    val cloudSolrClient = SolrSearch.getCloudSolrClient()
    val hBaseAdmin = new HBaseAdmin(hbaseConf)

    //创建hbase表
    if (!hBaseAdmin.isTableAvailable(tableName)) {
      val tableDesc = new HTableDescriptor(tableName)
      tableDesc.addFamily(new HColumnDescriptor(Bytes.toBytes("Basic")))
      tableDesc.addFamily(new HColumnDescriptor(Bytes.toBytes("OtherInfo")))
      hBaseAdmin.createTable(tableDesc)
    } else {
      print("Table " + tableName + " exists")
    }
    //索引名称
    val collectionName = propertie.getProperty("COLLECTION_NAME")
    //hbase表

    //查询所有索引名称
    val collectionNames = SolrSearch.queryAllCollections(cloudSolrClient)
    //判断要创建的索引名称是否已经存在,不存在则创建
    if (!collectionNames.contains(collectionName)) {
      SolrSearch.createCollection(collectionName)
    }
    //设置默认索引
    cloudSolrClient.setDefaultCollection(collectionName)
    kafkaStream.foreachRDD(rdd => {
      if (!rdd.isEmpty()) {
        rdd.map(x => (x.key(), x.value())).map(s => {
          val split = s._2.split("\\|")
          var put = new Put(Bytes.toBytes(split(1)))
          //将数据添加至hbase库
          put.addColumn(Bytes.toBytes("Basic"), Bytes.toBytes("name"), Bytes.toBytes(split(0)))
          put.addColumn(Bytes.toBytes("Basic"), Bytes.toBytes("age"), Bytes.toBytes(split(2)))
          put.addColumn(Bytes.toBytes("Basic"), Bytes.toBytes("name"), Bytes.toBytes(split(3)))
          put.addColumn(Bytes.toBytes("Basic"), Bytes.toBytes("sex"), Bytes.toBytes(split(4)))
          if (s._1.equals("hotel")) {
            //添加索引(姓名,地址,时间)
            SolrSearch.addDocs(cloudSolrClient, split(0), split(5), split(6), split(1))
            //将数据添加至hbase库
            put.addColumn(Bytes.toBytes("OtherInfo"), Bytes.toBytes("hotelAddr"), Bytes.toBytes(split(5)))
            put.addColumn(Bytes.toBytes("OtherInfo"), Bytes.toBytes("checkInTime"), Bytes.toBytes(split(6)))
            put.addColumn(Bytes.toBytes("OtherInfo"), Bytes.toBytes("checkOutTime"), Bytes.toBytes(split(7)))
            put.addColumn(Bytes.toBytes("OtherInfo"), Bytes.toBytes("acquaintance"), Bytes.toBytes(split(8)))
            (new ImmutableBytesWritable, put)
          } else if (s._1.equals("internet")) {
            //添加索引(姓名,地址,时间)
            SolrSearch.addDocs(cloudSolrClient, split(0), split(5), split(6), split(1))
            //将数据添加至hbase库
            put.addColumn(Bytes.toBytes("OtherInfo"), Bytes.toBytes("barAddr"), Bytes.toBytes(split(5)))
            put.addColumn(Bytes.toBytes("OtherInfo"), Bytes.toBytes("internetDate"), Bytes.toBytes(split(6)))
            put.addColumn(Bytes.toBytes("OtherInfo"), Bytes.toBytes("timeSpent"), Bytes.toBytes(split(7)))
            (new ImmutableBytesWritable, put)
          } else {
            //添加索引(姓名,地址,时间)
            SolrSearch.addDocs(cloudSolrClient, split(0), split(5), split(6), split(1))
            //将数据添加至hbase库
            put.addColumn(Bytes.toBytes("OtherInfo"), Bytes.toBytes("bayonetAddr"), Bytes.toBytes(split(5)))
            put.addColumn(Bytes.toBytes("OtherInfo"), Bytes.toBytes("checkDate"), Bytes.toBytes(split(6)))
            put.addColumn(Bytes.toBytes("OtherInfo"), Bytes.toBytes("tripType"), Bytes.toBytes(split(7)))
            (new ImmutableBytesWritable, put)
          }
        }).saveAsHadoopDataset(jobConf)
      }
    })
    ssc.start()
    ssc.awaitTermination()
  }

}

File: solutions/realtimeRetrieval/RealtimeRetrieval_java/src/main/java/com/huawei/bigdata/esandhbase/example/KafkaStreaming.java
package com.huawei.bigdata.esandhbase.example;
import com.huawei.bigdata.security.LoginUtil;
import com.huawei.bigdata.security.LoginUtilForKafka;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.hbase.HColumnDescriptor;
import org.apache.hadoop.hbase.HTableDescriptor;
import org.apache.hadoop.hbase.TableName;
import org.apache.hadoop.hbase.io.compress.Compression;
import org.apache.hadoop.hbase.io.encoding.DataBlockEncoding;
import org.apache.hadoop.hbase.util.Bytes;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.hadoop.hbase.client.*;
import org.apache.kafka.clients.consumer.ConsumerRecords;
import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.log4j.PropertyConfigurator;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.VoidFunction;
import org.apache.spark.streaming.Duration;
import org.apache.spark.streaming.api.java.JavaInputDStream;
import org.apache.spark.streaming.api.java.JavaStreamingContext;
import org.apache.spark.streaming.kafka010.*;
import org.elasticsearch.client.RestClient;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.apache.hadoop.hbase.HBaseConfiguration;

import java.io.FileInputStream;
import java.io.IOException;
import java.util.*;

import com.huawei.bigdata.security.LoginUtil;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.clients.consumer.ConsumerRecords;
import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.log4j.PropertyConfigurator;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.FileInputStream;
import java.util.Arrays;
import java.util.Properties;
/*
 * 执行producer之前，先参考shell/createTopic.sh创建topic，并配置权限。多个producer可以向一个topic写入。
 * 样例需要先启动KafkaStreaming，再启动DataProducer
 */
public class KafkaStreaming {
    static {
        PropertyConfigurator.configure(KafkaStreaming.class.getClassLoader().getResource("log4j.properties").getPath());
    }
    private static final Logger LOG = LoggerFactory.getLogger(KafkaStreaming.class);
    private static Properties properties = new Properties();

    private static TableName tableName = null;
    private static Connection conn = null;
    private static Admin admin = null;
    private static Table table = null;
    private static RestClient restClient = null;

    public static void main(String[] args)throws Exception {
        Configuration conf = HBaseConfiguration.create();
        //加载HDFS/HBase服务端配置，用于客户端与服务端对接
        conf.addResource(new Path(KafkaStreaming.class.getClassLoader().getResource("core-site.xml").getPath()));
        conf.addResource(new Path(KafkaStreaming.class.getClassLoader().getResource("hdfs-site.xml").getPath()));
        conf.addResource(new Path(KafkaStreaming.class.getClassLoader().getResource("hbase-site.xml").getPath()));
        //安全模式需要，普通模式可以删除s
        String krb5Conf =  KafkaStreaming.class.getClassLoader().getResource("krb5.conf").getPath();
        String keyTab = KafkaStreaming.class.getClassLoader().getResource("user.keytab").getPath();
        LoginUtilForKafka.setJaasFile("fwc", keyTab);
        LoginUtilForKafka.setKrb5Config(krb5Conf);
        LoginUtilForKafka.setZookeeperServerPrincipal("zookeeper/hadoop.hadoop.com");
        LoginUtil.login("fwc", keyTab, krb5Conf, conf);

        //加载consumer 配置 文件信息
        properties.load(new FileInputStream(KafkaStreaming.class.getClassLoader().getResource("consumer.properties").getPath()));
        String brokers = properties.getProperty("bootstrap.servers");
        String topics = properties.getProperty("topic");
        String batchTime = properties.getProperty("batchTime");
        String groupId = properties.getProperty("group.id");
        String indexName = properties.getProperty("indexName");

        //加载spark配置
        //创建一个Streaming启动环境。
        SparkConf sparkConf = new SparkConf().setAppName("KafkaToESAndHBase").setMaster("local[2]");
        sparkConf.set("spark.testing.memory", "2147480000");
        //streaming上下文，每隔多少秒执行一次获取批量数据然后处理这些数据
        JavaStreamingContext jsc = new JavaStreamingContext(sparkConf, new Duration(Long.parseLong(batchTime) * 1000));
        //为Streaming配置CheckPoint目录。它将足够多的信息checkpoint到某些具备容错性的存储系统如hdfs上，以便出错时能够迅速恢复
        jsc.checkpoint(properties.getProperty("checkPath"));

        //获取kafka使用的主题列表
        String[] topicArr = topics.split(",");
        Set<String> topicSet = new HashSet<String>(Arrays.asList(topicArr));
        Map<String, Object> kafkaParams = new HashMap();
        kafkaParams.put("bootstrap.servers", brokers);
        kafkaParams.put("value.deserializer", "org.apache.kafka.common.serialization.StringDeserializer");
        kafkaParams.put("key.deserializer", "org.apache.kafka.common.serialization.StringDeserializer");
        kafkaParams.put("group.id", groupId);
        kafkaParams.put("enable.auto.commit", "true");
        kafkaParams.put("auto.commit.interval.ms", "100"); //offset自动提交间隔

        // 配置kafka consumer的认证。运行kafkaUtils.createDirectStream代码在认证后才能消费数据
        kafkaParams.put("security.protocol","SASL_PLAINTEXT");
        kafkaParams.put("sasl.kerberos.service.name","kafka");

        //设置PreferConsistent方式，均匀分配分区。
        LocationStrategy locationStrategy = LocationStrategies.PreferConsistent();
        //对consumer进行自定义配置。Subscribe方法提交参数列表和kafka参数的处理。
        ConsumerStrategy consumerStrategy = ConsumerStrategies.Subscribe(topicSet, kafkaParams);

        ESSearch.init();
        restClient = ESSearch.getRestClient();
        //判断要创建的索引名称是否已经存在,不存在则创建
        if (!ESSearch.exist(restClient,indexName)) {
            ESSearch.createIndex(indexName);
        }

        try
        {
            conn = ConnectionFactory.createConnection(conf);
        }
        catch (Exception e)
        {
            LOG.error("Failed to createConnection because ", e);
        }
        tableName = TableName.valueOf(properties.getProperty("tableName"));
        table = conn.getTable(tableName);
        try {
            admin = conn.getAdmin();
            if (!admin.tableExists(tableName))
            {
                LOG.info("Creating table...");
                HTableDescriptor htd = new HTableDescriptor(tableName);
                HColumnDescriptor hcd1 = new HColumnDescriptor("Basic");
                HColumnDescriptor hcd2 = new HColumnDescriptor("OtherInfo");
                hcd1.setCompressionType(Compression.Algorithm.SNAPPY);
                hcd1.setDataBlockEncoding(DataBlockEncoding.FAST_DIFF);
                hcd2.setCompressionType(Compression.Algorithm.SNAPPY);
                hcd2.setDataBlockEncoding(DataBlockEncoding.FAST_DIFF);
                htd.addFamily(hcd1);
                htd.addFamily(hcd2);
                // 指定起止RowKey和region个数；此时的起始RowKey为第一个region的endKey，结束key为最后一个region的startKey。
                admin.createTable(htd, Bytes.toBytes(100000),   Bytes.toBytes(400000), 10);
            }
            else
            {
                LOG.warn("table already exists");
            }
        } catch (IOException e) {
            e.printStackTrace();
        }

        //从Kafka接收数据并生成相应的DStream( DStream操作最终会转换成底层的RDD的操作)
        JavaInputDStream<ConsumerRecord<String, String>> messages = KafkaUtils.createDirectStream(jsc, locationStrategy, consumerStrategy);
        messages.foreachRDD(
            new VoidFunction<JavaRDD<ConsumerRecord<String, String>>>() {
                @Override
                public void call(JavaRDD<ConsumerRecord<String, String>> consumerRecordJavaRDD) throws Exception {
                    //使用 foreachPartition 方法，每个分区并发工作
                    consumerRecordJavaRDD.foreachPartition(
                        new VoidFunction<Iterator<ConsumerRecord<String, String>>>() {
                             @Override
                             public void call(Iterator<ConsumerRecord<String, String>> consumerRecordIterator) throws Exception {
                                  hbaseAndESWrite(consumerRecordIterator);
                             }
                        }
                     );
                }
            }
        );
        // Spark Streaming系统启动
        jsc.start();
        jsc.awaitTermination();

        if (table != null) {
            try {
                table.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        if (admin != null) {
            try {
                admin.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        if (conn != null) {
            try {
                // Close the HBase connection.
                conn.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        //在进行完Elasticsearch操作后，需要调用“restClient.close()”关闭所申请的资源。
        if( restClient!=null) {
            try {
                restClient.close();
                LOG.info("Close the client successful in main.");
            } catch (Exception e1) {
                LOG.error("Close the client failed in main.",e1);
            }
        }
    };

    private static void hbaseAndESWrite(Iterator<ConsumerRecord<String, String>> consumerRecordIterator) throws IOException {
        try {
            List<Put> putList = new ArrayList<Put>();
            while (consumerRecordIterator.hasNext()) {
                ConsumerRecord cr = consumerRecordIterator.next();
                String key = cr.key().toString();
                String row = cr.value().toString();
                LOG.info("start Write data to ES and HBase.....");
                String[] elements = row.split(",");
                Put put = new Put(Bytes.toBytes(elements[1]));
                //将数据添加至hbase库
                put.addColumn(Bytes.toBytes("Basic"), Bytes.toBytes("name"), Bytes.toBytes(elements[0]));
                put.addColumn(Bytes.toBytes("Basic"), Bytes.toBytes("id"), Bytes.toBytes(elements[1]));
                put.addColumn(Bytes.toBytes("Basic"), Bytes.toBytes("age"), Bytes.toBytes(elements[2]));
                put.addColumn(Bytes.toBytes("Basic"), Bytes.toBytes("sex"), Bytes.toBytes(elements[3]));
                if(key.equals("hotel")){
                    //添加索引(姓名,地址,时间)
                    ESSearch.putData(restClient,elements[1], elements[0], elements[4], elements[5]);
                    //将数据添加至hbase库
                    put.addColumn(Bytes.toBytes("OtherInfo"), Bytes.toBytes("hotelAddr"), Bytes.toBytes(elements[4]));
                    put.addColumn(Bytes.toBytes("OtherInfo"), Bytes.toBytes("checkInTime"), Bytes.toBytes(elements[5]));
                    put.addColumn(Bytes.toBytes("OtherInfo"), Bytes.toBytes("checkOutTime"), Bytes.toBytes(elements[6]));
                    put.addColumn(Bytes.toBytes("OtherInfo"), Bytes.toBytes("acquaintance"), Bytes.toBytes(elements[7]));
                }else if(key.equals("internet")){
                    //添加索引(姓名,地址,时间)
                    ESSearch.putData(restClient,elements[1], elements[0], elements[4], elements[5]);
                    //将数据添加至hbase库
                    put.addColumn(Bytes.toBytes("OtherInfo"), Bytes.toBytes("barAddr"), Bytes.toBytes(elements[4]));
                    put.addColumn(Bytes.toBytes("OtherInfo"), Bytes.toBytes("internetDate"), Bytes.toBytes(elements[5]));
                    put.addColumn(Bytes.toBytes("OtherInfo"), Bytes.toBytes("timeSpent"), Bytes.toBytes(elements[6]));
                }else{
                    //添加索引(姓名,地址,时间)
                    ESSearch.putData(restClient,elements[1], elements[0], elements[4], elements[5]);
                    //将数据添加至hbase库
                    put.addColumn(Bytes.toBytes("OtherInfo"), Bytes.toBytes("bayonetAddr"), Bytes.toBytes(elements[4]));
                    put.addColumn(Bytes.toBytes("OtherInfo"), Bytes.toBytes("checkDate"), Bytes.toBytes(elements[5]));
                    put.addColumn(Bytes.toBytes("OtherInfo"), Bytes.toBytes("tripType"), Bytes.toBytes(elements[6]));
                }
                putList.add(put);
            }
            if (putList.size() > 0) {
                LOG.info("Write data to HBase.....");
                table.put(putList);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
